Exceptions in OMG frames don't get caught with wasmOMGOptimizationLevel=0

ioannad opened the pull request:
#### be01dbda2568b44142d4b3fc8f05a22b30ec1866
<pre>
Exceptions in OMG frames don&apos;t get caught with wasmOMGOptimizationLevel=0
<a href="https://bugs.webkit.org/show_bug.cgi?id=298651">https://bugs.webkit.org/show_bug.cgi?id=298651</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

When `--wasmOMGOptimizationLevel=0` the `PCToOriginMap`s do not get items appended,
as seen in `JSC::B3::Air::generate(...)`.
This results in `PCToCodeOriginMap`s without code ranges, therefore by construction they
have `m_pcRangeStart == m_pcRangeEnd == -1`.

When the PCToCodeOriginMap has no code ranges, and an exception is thrown from a call
surrounded by a try-catch_all, findPC can&apos;t find the code origin and the excepion
doesn&apos;t get caught.

This patch populates the PCToOriginMap when code.optLevel() = 0.
Since OMG frames are always expected to have their PCToOriginMap populated, this adds an assertion at
materializePCToOriginMap that maps are always populated with code ranges. The new test runs with the
flags that produced the bug, and includes an empty function, to make sure that even in this case,
there are origins in the PCToOriginMap.

Also add some optional logging to ensure the pc maps don&apos;t explode in size.

* JSTests/wasm/stress/exception-thrown-from-call.js: Added.
* JSTests/wasm/stress/exception-thrown-from-call.wasm: Added.
* JSTests/wasm/stress/exception-thrown-from-call.wat: Added.
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::GenerateAndAllocateRegisters::generate):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::materializePCToOriginMap):

Canonical link: <a href="https://commits.webkit.org/300252@main">https://commits.webkit.org/300252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a702752e98c615dbf2662b9870412f5687987ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128348 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73899 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50102 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92561 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61517 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33670 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109089 "Found 2 new API test failures: TestIPC.ConnectionTest/ConnectionRunLoopTest.SendAsyncAndInvalidate/ClientIsA, TestIPC.ConnectionTest/ConnectionRunLoopTest.SendAsyncAndInvalidate/ServerIsA (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73220 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32681 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27252 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71860 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113951 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103170 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131124 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120331 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37067 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101145 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49116 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105306 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101017 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25628 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46383 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24488 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45445 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48602 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54329 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150492 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48072 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38475 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51421 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e2190e8a1e5ec8bc72e09cadda6a75f1a353a969. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

Constellation reviewed (changes requested):
Nice catch!

We should populate PC map regardless of optimization levels. Thus that side needs to be fixed instead of here.

Also, can you attach a test?

mikhailramalho commented on Source/JavaScriptCore/interpreter/StackVisitor.cpp:
```diff
@@ -221,9 +221,17 @@ void StackVisitor::readInlinableNativeCalleeFrame(CallFrame* callFrame)
-        // While it can be pointing at the broken offset (e.g. all ARM64 instructions are 4-byte aligned), it is still fine since map is controlling pc with range.
-        auto callSiteIndexFromPC = omgCallee.tryGetCallSiteIndex(std::bit_cast<void*>(std::bit_cast<uintptr_t>(removeCodePtrTag<void*>(m_frame.m_returnPC)) - 1));
+        void* currentReturnPC = m_frame.m_returnPC;
+        std::optional<CallSiteIndex> callSiteIndexFromPC;
```
Nit: can you initialize callSiteIndexFromPC to nullptr instead of the else branch?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8b5600ec3c0ca4930c794fb679edca7c3205c27e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/29ccea1f648ee84f40de24adee5ce283dd10a6b0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/27ef7fcb6242a82762ebfb71c86db6371d3c81e6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/39a67f7782c142cb1d7a6e5219f2a96646999b61. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dbce41162d4c7f47c0b4b68fac79f97dccb93143. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a2200954d9750731adcd2b405d8d994e35b4b03f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/20a39a8f89cc75434f399410195adf694e799993. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50539

ioannad committed:
Exceptions in OMG frames don't get caught with wasmOMGOptimizationLevel=0
https://bugs.webkit.org/show_bug.cgi?id=298651

Reviewed by Yusuke Suzuki and Justin Michaud.

When `--wasmOMGOptimizationLevel=0` the `PCToOriginMap`s do not get items appended,
as seen in `JSC::B3::Air::generate(...)`.
This results in `PCToCodeOriginMap`s without code ranges, therefore by construction they
have `m_pcRangeStart == m_pcRangeEnd == -1`.

When the PCToCodeOriginMap has no code ranges, and an exception is thrown from a call
surrounded by a try-catch_all, findPC can't find the code origin and the excepion
doesn't get caught.

This patch populates the PCToOriginMap when code.optLevel() = 0.
Since OMG frames are always expected to have their PCToOriginMap populated, this adds an assertion at
materializePCToOriginMap that maps are always populated with code ranges. The new test runs with the
flags that produced the bug, and includes an empty function, to make sure that even in this case,
there are origins in the PCToOriginMap.

Also add some optional logging to ensure the pc maps don't explode in size.

* JSTests/wasm/stress/exception-thrown-from-call.js: Added.
* JSTests/wasm/stress/exception-thrown-from-call.wasm: Added.
* JSTests/wasm/stress/exception-thrown-from-call.wat: Added.
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::GenerateAndAllocateRegisters::generate):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::materializePCToOriginMap):

Canonical link: https://commits.webkit.org/300252@main
